### PR TITLE
[glslang] update to 14.0.0 / [shaderc] update to 2023.8

### PR DESCRIPTION
**Pull request by moritz-h**
- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md)
- [x] SHA512s are updated for each updated download
- [x] The "supports" clause reflects platforms that may be fixed by this new version
- [x] Any fixed [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt) entries are removed from that file.
- [x] Any patches that are no longer applied are deleted from the port's directory.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is added to each modified port's versions file.

### CODE REVIEW

**Comment by moritz-h** [comment]
Updating glslang seems becoming complicated with many thing breaking.
- The first issue is shaderc finding Python. Upstream version 2023.8 has support for glslang 14, but also the way how Python is found changed from `find_host_package(PythonInterp 3 REQUIRED)` to `find_package(Python COMPONENTS Interpreter REQUIRED)`. This does not find the Python version provided in the portfile anymore. I have tried `-DPython_EXECUTABLE=${PYTHON3}` with no success. As a quick testing workaround I made a patch to undo this upstream change. But this is probably not a nice solution.
- Second think broken is `vulkan-sdk-components`. It seems to use the CMake build in FindVulkan script with the glslang component. Unfortunately CMake is checking for every single glslang target, see https://github.com/Kitware/CMake/blob/v3.28.1/Modules/FindVulkan.cmake#L543, but the OGLCompiler target was removed with glslang 14, see release notes: https://github.com/KhronosGroup/glslang/releases/tag/14.0.0

Help or ideas how this should be fixed is highly appreciated.

**Comment by dg0yt** [comment]
The python problem is surprising, the pattern is fairly common. At least if there is no mistake as a lacking `vcpkg_find_acquire_program(PYTHON3)`.

vulkan-sdk-components really doesn't do much more than `find_package(Vulkan COMPONENTS glslang)`. This is a reasonable check, given the `glslang` dependency. Disabling the check just hides the problem. Users still couldn't use that command. (The problem exists independent of port vulkan-sdk-components.) As a mitigation, a vcpkg cmake wrapper for package `Vulkan` might perhaps add the missing target as an empty interface or as an alias. However, it is tricky: IMO the wrapper should be owned by port `vulkan`, but depends on the version of `glslang`...

**Comment by moritz-h** [comment]
Reported to CMake about the FindVulkan issue: https://gitlab.kitware.com/cmake/cmake/-/issues/25583

@dg0yt May it would be practical to just add that wrapper to the `glslang` port. It could be seen as a fix for the `find_package(Vulkan)` command, which would be an argument for fixing this in the `vulkan` port. But on the other hands all other regular targets for `find_package(Vulkan COMPONENTS glslang)` are also provided by the `glslang` port anyway.

And noted a fully independent discussion: Which versioning should glslang use? The current official releases or also switch to vulkan sdk version? I noted that all vulkan sdk related thinks switched recently to using the vulkan sdk version in #35479. glslang is part of the Vulkan SDK and also tags sdk versions in the repository. So may could argue to also use vulkan sdk versioning for glslang for consistency. On the other glslang uses own versioning for what upstream defines as release which is probably a strong argument to just keep current versioning. Further, depedencies such as shaderc seems to reference that official version schema.

**Comment by dg0yt** [comment]
> May it would be practical to just add that wrapper to the `glslang` port. It could be seen as a fix for the `find_package(Vulkan)` command, which would be an argument for fixing this in the `vulkan` port.

The problem is that it won't scale when the next component comes with a similar problem. Only one port can own the vulkan wrapper, and this responsibility belongs to the vulkan port. But it may chainload optional cmake scripts from the individual component ports.

> Which versioning should glslang use?

Switching version-scheme is bad, and glslang using glslang versioning is fine. 
IMO https://github.com/microsoft/vcpkg/pull/35479 had some aspects which weren't understood well enough before merge. I see the spirv-tools change, but at least it increases consistency within the spirv family.
And we can't enforce the same version between `vulkan`, `vulkan-sdk-components` and the actual component versions...

**Comment by moritz-h** [comment]
> IMO #35479 had some aspects which weren't understood well enough before merge. I see the spirv-tools change, but at least it increases consistency within the spirv family. And we can't enforce the same version between `vulkan`, `vulkan-sdk-components` and the actual component versions...

I start to understand your concerns. After looking at the FindVulkan CMake module trying to start a fix (https://gitlab.kitware.com/cmake/cmake/-/merge_requests/9158) I think the CMake module is named wrong and instead of being generically named `FindVulkan`, the better name would be `FindVulkanSDK`. It seems very specialized about seeing the Vulkan SDK as a single thing and not as collection of many individual libraries. If vcpkg wants to support the CMake built-in FindVulkan module it probably would be best to just install the SDK as one peace instead of trying to reproduce something similar from individual components.
On the other hand, most Vulkan related libraries (i.e. vulkan-headers, glslang, shaderc, ...) seems to install proper CMake config files. So it may be better to just advice users to use that individual CMake packages and ignore find_package(Vulkan) as a whole, because vcpkg simply does not provide the SDK but the single libraries to be used instead.
The only other thing I could think of is either fixing the upstream CMake module to be more flexible or fully overwrite it here by providing a VulkanConfig.cmake which maybe has a similar/compatible interface but is just wrapping all targets from the individual libraries. But that sounds not nice to maintain.

**Comment by moritz-h** [comment]
(@dg0yt @Neumann-A ping as you recently updated the Vulkan things)

FYI: CMake will remove the FindVulkan module via a policy, see https://gitlab.kitware.com/cmake/cmake/-/issues/25617
All Vulkan related libraries seems to now have CMake config files in the upstream libraries. And maybe even the next Vulkan SDK release starts shipping CMake config files (at least found [a confirmation for glslang](https://vulkan.lunarg.com/issue/view/656df8aa5df1125b58afb491), but hopefully they include it directly for all libraries)

As FindVulkan was historically written specific to the Vulkan SDK Release and never with installing all Vulkan libraries individually in mind, vcpkg could may just drop support for the FindVulkan module. Maybe would make sense to wait until the next Vulkan SDK release is tagged, then all libraries could be updated here, and then tell users to use the new official way by directly using the cmake configs of the libraries and ignore the FindVulkan module.

**Comment by Neumann-A** [comment]
Evennif CMake deprecates it FindVulkan still needs to be supported by vcpkg due to not everybody updating to latest cmake

**Comment by moritz-h** [comment]
That is not exactly what I meant. I meant drop support in the sense of not actively fixing in vcpkg what is broken already upstream and maybe remove actively testing find_package(Vulkan). Also the suggested replacement (using library cmake config files directly) works for all CMake versions.

Also I noted the new SDK version seems released now. So installing upstream Vulkan SDK 1.3.275.0 would be equivalent to merging this PR.

Maybe just look at all possible cases:

- Vulkan SDK <= 1.3.268.0, CMake <= 3.28
  - find_package(Vulkan) works

- Vulkan SDK <= 1.3.268.0, CMake >= 3.29
  - find_package(Vulkan) works, if CMP0160 is OLD

- Vulkan SDK >= 1.3.275.0, CMake <= 3.28
  - find_packge(Vulkan COMPONENT glslang) is broken, exactly what is happening in this PR, rest of find_package(Vulkan) probably still works until further breaking changes in the Vulkan SDK happen.
  - I'm suggesting to remove the find_package(Vulkan) check from vcpkg and not try to fix what is broken already upstream. Therefore also not adding any wrapper to make glslang 14.0.0 still work with find_package(Vulkan COMPONENT glslang).
  - As a replacement user should migrate to use the config files provided from the Vulkan libs, which are now also included in the Vulkan SDK release. Switching to this can be done fully independant from the CMake version, and is probably the modern way intended by Vulkan and CMake developers.

- Vulkan SDK >= 1.3.275.0, CMake >= 3.29
  - Users should migrate to use config files provided from the vulkan libs
  - find_package(Vulkan) is deprecated in CMake and at least the glslang component is broken, should not be used anymore (and therefore not required to work by vcpkg)

**Comment by Neumann-A** [comment]
> I'm suggesting to remove the find_package(Vulkan) check from vcpkg and not try to fix what is broken already upstream. Therefore also not adding any wrapper to make glslang 14.0.0 still work with find_package(Vulkan COMPONENT glslang).

The problem is it breaks ports using that call. I rather see the call succeed than have to patch all downstream consumers.

**Comment by moritz-h** [comment]
The thing is, it would break anyway if ports would use the latest upstream Vulkan SDK. So far I thought one of the vcpkg principles is to follow upstream as close as posible. I just would see this whole process as breaking change of Vulkan. Sooner or later all projects are required to migrate to the new way.

But practically speaking, it probably takes a long time until all projects have migrated by themself, and you are right probably until then patching many ports is not practical. Maybe as short term solution as currently only find_package(Vulkan COMPONENT glslang) is broken and not the whole find_package(Vulkan) call, the vulkan-sdk-components test should just remove the check for the glslang component and keep the rest. Otherwise it will not be possible to update glslang.

**Comment by moritz-h** [comment]
Now with #36262 merged this could be seen as the remaining part of the vulkan SDK update. glslang 14.0.0 is the same version included in the 1.3.275.0 vulkan SDK release.

File: ports/glslang/usage [hunk 1]
```diff
@@ -1,4 +1,4 @@
 glslang provides CMake targets:
 
     find_package(glslang CONFIG REQUIRED)
-    target_link_libraries(main PRIVATE glslang::OSDependent glslang::glslang glslang::MachineIndependent glslang::GenericCodeGen glslang::glslang-default-resource-limits glslang::OGLCompiler glslang::SPVRemapper glslang::SPIRV glslang::HLSL)
+    target_link_libraries(main PRIVATE glslang::glslang glslang::glslang-default-resource-limits glslang::GenericCodeGen glslang::MachineIndependent glslang::OSDependent glslang::SPIRV glslang::SPVRemapper)
```
Comment (dg0yt) [inline ports/glslang/usage:4]: Are user really mean to link more than `glslang::glslang`?

Comment (moritz-h) [reply ports/glslang/usage:4]: At least some. But now that you mention it, `GenericCodeGen` `MachineIndependent` and `OSDependent` should be removed here ([source](https://github.com/KhronosGroup/glslang/issues/3462#issuecomment-1871369876))
I know that `glslang-default-resource-limits` is sometimes needed from own projects. During the `vsg` patch cleanup I thought also `SPIRV` is required. No idea about `SPVRemapper`.

**Comment by LilyWangLL** [comment]
Please disable option `SHADERC_ENABLE_COPYRIGHT_CHECK` for `shaderc`, the following codes will open Visual Studio Code in building progress.
https://github.com/google/shaderc/blob/9bd299b0f28ba53308567cc17d00fb1bfe93e2c3/CMakeLists.txt#L100-L106
```
if (SHADERC_ENABLE_COPYRIGHT_CHECK)
  add_custom_target(check-copyright ALL
    ${PYTHON_EXECUTABLE} ${CMAKE_CURRENT_SOURCE_DIR}/utils/add_copyright.py
    --check
    WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
    COMMENT "Check copyright")
endif()
```

**Comment by LilyWangLL** [comment]
All features passed with following triplets:
```
x86-windows
x64-windows
x64-windows-static
```
Usage test passed on `x64-windows`.

**Comment by moritz-h** [comment]
@Neumann-A 
CMake 3.29.0 will fix the broken FindVulkan module, see https://gitlab.kitware.com/cmake/cmake/-/merge_requests/9225
I added a CMake version check around testing for the glslang component. So if using CMake >= 3.29 nothing changes and for CMake <= 3.28 the check for the glslang component is skiped in `vulkan-sdk-components`. Are you also fine with this solution then?

For additional context, CMake and Vulkan team are working together to fully deprecate the FindVulkan module in future, and Vulkan components provide correct CMake configs directly, see https://gitlab.kitware.com/cmake/cmake/-/issues/25617#note_1480078. Currently planned for CMake 3.30.

@LilyWangLL 
What do you mean with open VS Code? The code you linked executes a python script. I do not know how this is related to VS Code. While the exact code snippet you linked is just a check that probably could be skipped, the same code is probably required right below, where the same python script is executed: https://github.com/google/shaderc/blob/v2023.8/CMakeLists.txt#L108-L111

**Comment by LilyWangLL** [comment]
> @LilyWangLL do you have an update on what exactly the problem with "VS code" is?

If you installed Visual Studio Code on your machine, the installation will open it.

**Comment by moritz-h** [comment]
Ok, I think I found something. This has probably nothing to do with VS Code itself, but you configured your system that it opens Python files with VS Code instead of executing them.
Nevertheless, that shaderc tries to execute the Python scripts directly instead of explicitly calling the Python executable is probably a bug. I have submitted an upstream fix and updated the fix-python patch here. For technical details see upstream PR: https://github.com/google/shaderc/pull/1401
Can you please try again now.

**Review by LilyWangLL — APPROVED** [review]

**Review by BillyONeal — APPROVED** [review]
Next update I think we should avoid making build-version.inc need to be manually updated with this:

```diff
diff --git a/ports/shaderc/build-version.inc b/ports/shaderc/build-version.inc
index 35c110123..a793b27fa 100644
--- a/ports/shaderc/build-version.inc
+++ b/ports/shaderc/build-version.inc
@@ -1 +1 @@
-"shaderc v2023.8 v2023.8\n"
+"shaderc v@VERSION@ v@VERSION@\n"
diff --git a/ports/shaderc/portfile.cmake b/ports/shaderc/portfile.cmake
index 45a2ec300..f09349c79 100644
--- a/ports/shaderc/portfile.cmake
+++ b/ports/shaderc/portfile.cmake
@@ -14,7 +14,7 @@ vcpkg_from_github(
         fix-python.patch # Upstream PRs #1389 and #1401.
 )
 
-file(COPY "${CMAKE_CURRENT_LIST_DIR}/build-version.inc" DESTINATION "${SOURCE_PATH}/glslc/src")
+configure_file("${CMAKE_CURRENT_LIST_DIR}/build-version.inc" "${SOURCE_PATH}/glslc/src/build-version.inc" @ONLY LF)
 
 set(OPTIONS "")
 if(VCPKG_CRT_LINKAGE STREQUAL "dynamic")
```

but it isn't worth blocking an otherwise working build to get that.

**Review by vicroms — APPROVED** [review]

**Comment by BillyONeal** [comment]
Thanks for the updates!

**Pull request merged**